utils/tfw_logger.cpp

```diff
@@ -0,0 +1,214 @@
+
+	auto col = tfw_column_factory->create(Type::UInt64);
+	block->AppendColumn("timestamp", col);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		const TfwField *field = &tfw_fields[i];
+
+		auto col = tfw_column_factory->create(field->code);
+		block->AppendColumn(field->name, col);
+	}
+
+	return block;
+}
+
+int
+read_access_log_event(const char *data, int size, int cpu)
+{
+	Block *block = tfw_clickhouse->getBlock(cpu);
+	const char *p = data;
+	TfwBinLogEvent *event = (TfwBinLogEvent *)p;
+	int i;
+
+	p += sizeof(TfwBinLogEvent);
+	size -= sizeof(TfwBinLogEvent);
+
+#define INT_CASE(method, col_type, val_type)					\
+	case method:								\
+		if (len > size)							\
+			return -1;						\
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i))			\
+			(*block)[ind]->As<col_type>()->Append(*(val_type *)p);	\
+		else								\
+			(*block)[ind]->As<col_type>()->Append(0);		\
+		break;
+
+
+	(*block)[0]->As<ColumnUInt64>()->Append(event->timestamp);
+
+	for (i = TFW_MMAP_LOG_ADDR; i < TFW_MMAP_LOG_MAX; ++i) {
+		int len, ind = i + 1;
+
+		len = tfw_mmap_log_field_len((TfwBinLogFields)i);
+
+		switch (i) {
+		INT_CASE(TFW_MMAP_LOG_ADDR, ColumnIPv6, struct in6_addr);
+		INT_CASE(TFW_MMAP_LOG_METHOD, ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_VERSION, ColumnUInt8, unsigned char);
+		INT_CASE(TFW_MMAP_LOG_STATUS, ColumnUInt16, uint16_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_CONT_LEN, ColumnUInt32, uint32_t);
+		INT_CASE(TFW_MMAP_LOG_RESP_TIME, ColumnUInt32, uint32_t);
+
+		case TFW_MMAP_LOG_VHOST:
+		case TFW_MMAP_LOG_URI:
+		case TFW_MMAP_LOG_REFERER:
+		case TFW_MMAP_LOG_USER_AGENT:
+			if (!TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+				(*block)[ind]->As<ColumnString>()->Append(move(string("")));
+				break;
+			}
+			len = *((uint16_t *)p);
+			if (len + 2 > size)
+				return -1;
+			(*block)[ind]->As<ColumnString>()->Append(move(string(p + 2, len)));
+			len += 2;
+			break;
+		default:
+			cerr << "Unknown field type: " << i << endl;
+			return -1;
+		}
+
+		if (TFW_MMAP_LOG_FIELD_IS_SET(event, i)) {
+			p += len;
+			size -= len;
+		}
+	}
+
+	return p - data;
+}
+
+void
+callback(const char *data, int size, unsigned int cpu)
+{
+	TfwBinLogEvent *event;
+	const char *p = data;
+	int r;
+
+#ifdef DEBUG
+	hexdump(data, size);
+#endif /* DEBUG */
+
+	do {
+		if (size < (int)sizeof(TfwBinLogEvent))
+			return;
+
+		event = (TfwBinLogEvent *)p;
+
+		switch (event->type) {
+		case TFW_MMAP_LOG_TYPE_ACCESS:
+			r = read_access_log_event(p, size, cpu);
+			if (r < 0)
+				return;
+			size -= r;
+			p += r;
+			break;
+		case TFW_MMAP_LOG_TYPE_DROPPED:
+			p += sizeof(TfwBinLogEvent);
+			size -= sizeof(TfwBinLogEvent);
+			if ((int)sizeof(u64) > size) {
+				cerr << "Incorrect event length" << endl;
+				return;
+			}
+
+			cerr << "Dropped events: " << *(u64 *)p << endl;
+			size -= sizeof(u64);
+			p += sizeof(u64);
+			return;
+		default:
+			cerr << "Unsupported log type: " << event->type << endl;
+			return;
+		}
+	} while (size > 0);
```

Comment:
I'm not sure if I understood what's going on: we should have fixed size of Clickhouse batches, but I don't see the limit. It seems we collect a batch untill we have something in the RB, and this may be for infinite time (softirq is more prioritized than the user space threads!).

Reply:
This callback is called with data limited by the buffer size. Inside the loop we don't move the tail pointer. It occurs only at the end of the callback:
`tfw_clickhouse->commit(cpu);`
If such limitation is not enough, I could add an events counter to this loop.